android/app/src/main/java/com/zulipmobile/MainApplication.java

```diff
@@ -85,11 +85,13 @@ public IPushNotification getPushNotification(Context context, Bundle bundle, App
             clearConversations(conversations);
             NotificationManager nMgr = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
             nMgr.cancelAll();
-            return null;
         } else {
             PushNotificationsProp prop = new PushNotificationsProp(bundle);
-            addConversationToMap(prop, conversations);
-            return new GCMPushNotifications(context, bundle, defaultFacade, defaultAppLaunchHelper, new JsIOHelper(), conversations);
+            if (prop.getEvent().equals("message")) {
+                addConversationToMap(prop, conversations);
+                return new GCMPushNotifications(context, bundle, defaultFacade, defaultAppLaunchHelper, new JsIOHelper(), conversations);
+            }
         }
+        return null;
```

Comment:
The main question I have on looking at this code is:

Hmm. Why is this returning null? What's the semantics of `getPushNotification` returning null?

I ended up with a lot to say and decided to make it visible for the whole team, so I went into detail [in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/subject/android.20notifications/near/626529). Please go ahead and read that discussion if you haven't already.

For the record here, the bottom line for this PR:
* Returning null where we currently do is OK.
* Returning null in the GCM case isn't right. But! We can instead throw an `IPushNotification.InvalidNotificationException`, and that will have the right effect. Let's do that for the immediate fix here.

Reply:
Throwing the exception is not straightforward here, as the `get` method in `PushNotification.java` is not handling the Exception, hence I'll see for some workarounds so we don't have to return null.

Reply:
I see, yeah.

Checked exceptions -- [not one of Java's better decisions](https://www.javaworld.com/article/3142626/core-java/are-checked-exceptions-good-or-bad.html).

I think the right way to do this in the API of this library is to have the `onReceive` on our `IPushNotification` implementation take its own look at things and throw that exception instead.

... Ha, wait, actually, instead of throwing the exception, that method can just... not do anything. This exception type is actually completely meaningless. The one method that has the option to throw it, can just return instead and get the same effect.

So yeah. I think we can just write
```
@Override
public void onReceived() {
  if (/* ... */) {
    return;
  }
  super.onReceived();
}
```
and that'll do the job.

Reply:
Another way can be, I'll implement my own GcmMessageHandlerService and handle the null returns?
I'll have to patch the AndroidManifest though.

Reply:
This AndroidManifest thing is not possible as it doesn't allow changing the attr `android:name` and that is the unique key as well.